design-documents/asynchronous-operations/consumer-processes-improvements.md

```diff
@@ -0,0 +1,89 @@
+
+[MC-19250](https://github.com/magento/magento2/commit/269b47af3e37fbbe76e9f38d45fdb0cf969d45e3) was recently introduced in the code base. ([Docs](https://github.com/magento/devdocs/pull/5289))  
+Which introduces an option `consumers_wait_for_messages`. When that option is set to false, a consumer will stop the moment it doesn't find any new messages.
+
+The problem with this option, is that every time the cronjob `consumers_runner` runs, it will spawn a new consumer process, the consumer checks if messages are available and if none found it will kill itself. So this means it will spawn unneeded processes which take up memory, live for a very short period and then disappear again.  
+
+In this proposal, the flag `only-spawn-when-message-available` would run logic to check *before* the consumer process gets spawned and not *after*.
+
+## Problem 2: deployment problems
+
+### The problem
+
+When deploying a new version of Magento/third party/custom code to a server, we can potentially run against a problem where consumers are still using old code loaded in memory. This is not only the code the consumer is running itself, but might also be Magento core code itself for the consumer processes themselves (which might change when upgrading Magento to a newer version).
+
+Also when using a capistrano-style deployment where you have a symlinked a `current` directory pointing to a certain release directory, after a new deploy the running consumers will still supposedly run from the old release directory referencing a `.pid` file containing its process id in a directory which is in the old release directory. The cronjob `consumers_runner` will go searching for that file in the new release directory and won't find it. Causing a second consumer to start up, even though the old one is still running. (This is probably already fixed by this unreleased commit: [MC-18477](https://github.com/magento/magento2/commit/1d9e07b218c7c8ad1f05706828cb2dd47d2d2d58))
+
+### The suggested solution
+
+The suggestion here would be to give deployment scripts the option to signal the running consumer proccesses to handle their current messages and then kill themselves as soon as possible.  
+We already have the poison pill functionality which could be used here, but the problem with that one is that it will only be checked when a new message appears in the queue. If no messages appear in the queue during a deploy, and we only update the poison pill version, the consumer will still keep running because it doesn't check for the poison pill version until a new message appears.  
+So I would suggest some kind of dummy message to be created in each queue which only purpose is for the consumers to check the poison pill version and that message then should get removed again from the queue and nothing should be done with it.
+
+I'm seeing this as some new command being added to `bin/magento` (`queue:consumers:suicide` ?) which does both these things:
```

Comment:
Probably, it's better to don't use `suicide` word here. Here is some interesting proposal for Solidity (a lang to create smart contracts on Etherium blockchain) where suicide() method existed from the very beginning:
https://github.com/Souptacular/EIPs/blob/55b2dcaea2526bfbb59c39930db8a13bb6c1357c/EIPS/eip-6.md
They deprecated it referring to the mental health of people working with the lang.

Reply:
Very good remark @roma-glushko! I should have thought of that.